File: packages/termworld/example/linux/test/text_input_boundary_test.cc
#include <cassert>
#include <string>
#include <vector>

int main() {
  const std::vector<std::pair<std::string, bool>> updates = {
      {u8"ㅎ", true}, {u8"한글", true}, {u8"韓國", true}, {u8"韓國", false}};
  std::vector<std::string> commits;
  for (const auto& update : updates) {
    if (!update.second) commits.push_back(update.first);
  }
  assert(commits.size() == 1);
  assert(commits.front() == u8"韓國");
  assert(std::string(u8"👩🏽‍💻").size() == 15);
  return 0;
}

File: packages/termworld/example/android/app/src/test/kotlin/com/example/termworld_example/TextInputBoundaryTest.kt
package com.example.termworld_example

import java.text.BreakIterator
import java.util.Locale
import org.junit.Assert.assertEquals
import org.junit.Test

class TextInputBoundaryTest {
    @Test
    fun preservesUtf16AndExtendedGraphemeBoundaries() {
        val value = "한글👩🏽‍💻e\u0301"
        val iterator = BreakIterator.getCharacterInstance(Locale.ROOT)
        iterator.setText(value)
        val boundaries = mutableListOf<Int>()
        var boundary = iterator.first()
        while (boundary != BreakIterator.DONE) {
            boundaries.add(boundary)
            boundary = iterator.next()
        }

        assertEquals(0, boundaries.first())
        assertEquals(value.length, boundaries.last())
        assertEquals(value, String(value.toCharArray()))
    }

    @Test
    fun composingReplacementCommitsOnlyFinalValue() {
        val commits = mutableListOf<String>()
        val updates = listOf("ㅎ" to true, "한글" to true, "韓國" to true, "韓國" to false)
        updates.filterNot { it.second }.forEach { commits.add(it.first) }

        assertEquals(listOf("韓國"), commits)
    }
}
